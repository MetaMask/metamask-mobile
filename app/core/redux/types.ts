import { AnyAction, Store } from 'redux';
import { RootState } from '../../reducers';

/**
 * Redux store type
 * TODO: Replace AnyAction with union type of all actions
 */
export type ReduxStore = Store<RootState, AnyAction>;
