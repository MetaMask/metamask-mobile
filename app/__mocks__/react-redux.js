export * from 'react-redux-test';
