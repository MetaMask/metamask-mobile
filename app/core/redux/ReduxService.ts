import Logger from '../../util/Logger';
import { ReduxStore } from './types';

/**
 * ReduxService class that manages the Redux store
 */
class ReduxService {
  static #reduxStore: ReduxStore;

  static #assertReduxStoreType(store: ReduxStore) {
    if (
      typeof store.dispatch !== 'function' ||
      typeof store.getState !== 'function'
    ) {
      const error = new Error('Redux store is not a valid store!');
      Logger.error(error);
      throw error;
    }
    return this.#reduxStore;
  }

  static #assertReduxStoreExists() {
    if (!this.#reduxStore) {
      const error = new Error('Redux store does not exist!');
      Logger.error(error);
      throw error;
    }
    return this.#reduxStore;
  }

  /**
   * Set the store in the Redux class
   * @param store
   */
  static set store(store: ReduxStore) {
    this.#assertReduxStoreType(store);
    this.#reduxStore = store;
  }

  /**
   * Get the store from the Redux class
   */
  static get store() {
    this.#assertReduxStoreExists();
    return this.#reduxStore;
  }
}

export default ReduxService;
