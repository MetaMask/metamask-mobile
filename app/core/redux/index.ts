export { default } from './ReduxService';

export * from './types';
