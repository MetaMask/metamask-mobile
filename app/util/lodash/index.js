"use strict";
exports.__esModule = true;
exports.isString = exports.gt = exports.lt = exports.gte = exports.lte = exports.isZero = void 0;
/**
 * Function that checks if value is zero
 *
 * @param value number | any
 * @returns
 */
// TODO: Replace "any" with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
var isZero = function (value) {
    var _c;
    if ((value === null || value === void 0 ? void 0 : value.toString) && ((_c = value === null || value === void 0 ? void 0 : value.toString) === null || _c === void 0 ? void 0 : _c.call(value)) === '0') {
        return true;
    }
    return false;
};
exports.isZero = isZero;
/**
 * Function that checks if value is less than or equal to other
 *
 * @param value number
 * @param other number
 * @returns
 */
var lte = function (value, other) {
    return Number(value) <= Number(other);
};
exports.lte = lte;
/**
 * Function that checks if value is greater than or equal to other
 *
 * @param value number
 * @param other number
 * @returns
 */
var gte = function (value, other) {
    return Number(value) >= Number(other);
};
exports.gte = gte;
/**
 * Function that checks if value is less than other
 *
 * @param value number
 * @param other number
 * @returns
 */
var lt = function (value, other) {
    return Number(value) < Number(other);
};
exports.lt = lt;
/**
 * Function that checks if value is greater than other
 *
 * @param value number
 * @param other number
 * @returns
 */
var gt = function (value, other) {
    return Number(value) > Number(other);
};
exports.gt = gt;
/**
 * Gets the `toStringTag` of `value`.
 *
 * @param {*} value The value to query.
 * @returns {string} Returns the `toStringTag`.
 */
// TODO: Replace "any" with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function getTag(value) {
    var objToString = Object.prototype.toString;
    if (value === null) {
        return value === undefined ? '[object Undefined]' : '[object Null]';
    }
    return objToString.call(value);
}
/**
 * Checks if `value` is classified as a `String` primitive or object.
 *
 * @param {*} value The value to check.
 * @returns {boolean} Returns `true` if `value` is a string, else `false`.
 */
// TODO: Replace "any" with type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isString(value) {
    var type = typeof value;
    return (type === 'string' ||
        (type === 'object' &&
            value !== null &&
            !Array.isArray(value) &&
            getTag(value) === '[object String]'));
}
exports.isString = isString;
