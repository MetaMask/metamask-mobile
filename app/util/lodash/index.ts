/**
 * Function that checks if value is zero
 *
 * @param value number | any
 * @returns
 */
export const isZero = (value: number | any): boolean => {
  if (value?.toString && value?.toString?.() === '0') {
    return true;
  }
  return false;
};

/**
 * Function that checks if value is less than or equal to other
 *
 * @param value number
 * @param other number
 * @returns
 */
export const lte = (value: number, other: number): boolean =>
  Number(value) <= Number(other);

/**
 * Function that checks if value is greater than or equal to other
 *
 * @param value number
 * @param other number
 * @returns
 */
export const gte = (value: number, other: number): boolean =>
  Number(value) >= Number(other);

/**
 * Function that checks if value is less than other
 *
 * @param value number
 * @param other number
 * @returns
 */
export const lt = (value: number, other: number): boolean =>
  Number(value) < Number(other);

/**
 * Function that checks if value is greater than other
 *
 * @param value number
 * @param other number
 * @returns
 */
export const gt = (value: number, other: number): boolean =>
  Number(value) > Number(other);
