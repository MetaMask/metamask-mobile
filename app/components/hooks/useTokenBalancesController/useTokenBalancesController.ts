import { useSelector } from 'react-redux';
import { isEqual } from 'lodash';
import { ControllerHookType } from '../controllerHook.types';
import { BN } from 'ethereumjs-util';
import { selectContractBalances } from '../../../selectors/tokenBalancesController';

interface TokenBalances {
  [address: string]: BN;
}

const useTokenBalancesController = (): ControllerHookType<TokenBalances> => {
  const tokenBalances = useSelector(selectContractBalances, isEqual);

  return { data: tokenBalances };
};

export default useTokenBalancesController;
export type { TokenBalances };
