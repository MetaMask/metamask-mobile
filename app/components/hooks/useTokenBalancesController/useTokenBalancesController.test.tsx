import React, { useEffect } from 'react';
import { Text } from 'react-native';
import { Provider } from 'react-redux';
import { createStore, Store } from 'redux';
import { act, render, waitFor } from '@testing-library/react-native';
import useTokenBalancesController from './useTokenBalancesController';
import { BN } from 'ethereumjs-util';
import { cloneDeep } from 'lodash';

// initial state for the test store
const initialState = {
  engine: {
    backgroundState: {
      TokenBalancesController: {
        contractBalances: {
          '0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95': new BN(0x2a),
        },
      },
    },
  },
};

// test reducer for the test store
const testBalancesReducer = (state: any, action: any) => {
  if (action.type === 'add-balances') {
    return {
      ...state,
      engine: {
        ...state.engine,
        backgroundState: {
          ...state.engine.backgroundState,
          TokenBalancesController: {
            ...state.engine.backgroundState.TokenBalancesController,
            contractBalances: {
              ...state.engine.backgroundState.TokenBalancesController
                .contractBalances,
              ...action.value,
            },
          },
        },
      },
    };
  }
  return state;
};

const spyOnDummyUseEffect = jest.fn();
// Dummy component to test the useTokenBalancesController hook
// and be able to detect the rerenders using the useEffect spy
const DummyTestComponent = () => {
  const balances = useTokenBalancesController();

  useEffect(() => {
    spyOnDummyUseEffect();
  }, [balances]);

  // render the balances as a string to be able to compare the snapshots
  return <Text>{JSON.stringify(balances)}</Text>;
};

describe('useTokenBalancesController()', () => {
  let toJSON: any;
  let testStore: Store<any, any>;

  beforeEach(async () => {
    // create a test store, not a mock, as we need to test for content changes
    // Cloning as the store is mutated by the reducer and prevents the test from
    // starting from fresh store with the initial state
    testStore = createStore(testBalancesReducer, cloneDeep(initialState));

    // console.log('testStore', JSON.stringify(testStore.getState()));

    const Wrapper = ({ children }: any) => (
      <Provider store={testStore}>{children}</Provider>
    );

    // render the component with the test store
    toJSON = render(<DummyTestComponent />, { wrapper: Wrapper }).toJSON;

    await waitFor(async () => {
      // check that the initial balances are returned
      expect(spyOnDummyUseEffect).toHaveBeenCalledTimes(1);
      expect(toJSON()).toMatchSnapshot();
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should rerender when state changed', async () => {
    // add a new balance to the store
    act(() => {
      testStore.dispatch({
        type: 'add-balances',
        value: {
          '0x326836cc6cd09B5aa59B81A7F72F25FcC0136b96': new BN(0x539),
        },
      });
    });

    // check that the useEffect is called again and the new updated balances are returned
    await waitFor(() => {
      expect(spyOnDummyUseEffect).toHaveBeenCalledTimes(2);
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it('should not rerender when state is not changed', async () => {
    // add the same balance to the store again, should override the previous one
    act(() => {
      testStore.dispatch({
        type: 'add-balances',
        value: {
          '0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95': new BN(0x2a),
        },
      });
    });

    // check that the useEffect is not called again and the returned balances are unchanged
    await waitFor(async () => {
      expect(spyOnDummyUseEffect).toHaveBeenCalledTimes(1);
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
