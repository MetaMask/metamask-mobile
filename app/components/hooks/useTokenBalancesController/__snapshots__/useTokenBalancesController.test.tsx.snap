// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTokenBalancesController() should not rerender when state is not changed 1`] = `
<Text>
  {"data":{"0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95":"2a"}}
</Text>
`;

exports[`useTokenBalancesController() should not rerender when state is not changed 2`] = `
<Text>
  {"data":{"0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95":"2a"}}
</Text>
`;

exports[`useTokenBalancesController() should rerender when state changed 1`] = `
<Text>
  {"data":{"0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95":"2a"}}
</Text>
`;

exports[`useTokenBalancesController() should rerender when state changed 2`] = `
<Text>
  {"data":{"0x326836cc6cd09B5aa59B81A7F72F25FcC0136b95":"2a","0x326836cc6cd09B5aa59B81A7F72F25FcC0136b96":"539"}}
</Text>
`;
